### ISSUE TITLE
lack of documentation about how to create a password rule

### ISSUE THREAD

**Issue by shdf2** [openbullet/openbullet.github.io#2]
Hello,
could you please make some documentation about how to create a password rule ?
let's say i want my passwords to be :
_Minimum  8 characters
Maximum 32 characters
with Digits
with uppercase
with lowercase_

how i do that ?

Thank you very much for your help

**Comment by openbullet** [comment]
Alright I'll see where to put more info about that, but anyways let's say your password slice is called PASSWORD. You will need to add rules like
`PASSWORD MinLength 8`
`PASSWORD MaxLength 32`
`PASSWORD Contains Digit`
`PASSWORD Contains Lowercase`
`PASSWORD Contains Uppercase`
If these are all verified then your password will be accepted as valid and it will be checked with the config, otherwise it will be immediately flagged as a FAIL and OB will move to the next one.

**Comment by shdf2** [comment]
Like this ?

![](https://i.imgur.com/no5GEOj.png)

**Comment by MeowBullet** [comment]
@shdf2 it depends what is your slice named in environment.ini? 

example below
```
[WLTYPE]
Name=Credentials
Regex=^.*:.*$
Verify=True
Separator=:
Slices=USER,PASS
``` 
so if your slice is called Pass1 which would for example be
```
Slices=User1,Pass1
```
If it however was PASSWORD
```
Slices=USERNAME,PASSWORD
```
Then you would do it like this 
````
PASSWORD MinLength 8
PASSWORD MaxLength 32
PASSWORD Contains Digit
PASSWORD Contains Lowercase
PASSWORD Contains Uppercase
```

**Comment by openbullet** [comment]
Thank you @MeowBullet for the explanation.

**Comment by openbullet** [comment]
Actually let me reopen so I remember to update the documentation with a section on this ^_^

**Comment by shdf2** [comment]
i have to study that thing because i really don't understand what is that "Slice" thing and its meaning...

The way to do it in BB what very clear, i don't understand why it is made that way now ?

Thanks @MeowBullet i'll try to understand...

**Comment by MeowBullet** [comment]
Go to your settings folder, open it up locate the environment.ini file, edit It to change your slice under credential to whatever it is you desire. The default is USERNAME, PASSWORD, However others including me reverted to USER,PASS.

The reason it's like this is its opened a big door of possibilities. It's handy to have the environment file and many of us like it.

**Comment by shdf2** [comment]
OK ! After reading your explanation 36 times i understood !!
in my environnement.ini i have this : 

> [WLTYPE]
> Name=Credentials
> Regex=^.*:.*$
> Verify=True
> Separator=:
> Slices=USERNAME,PASSWORD

i loaded a config made under BB 2.5.1 and converted to loli. and this is what i can see once loaded : 
![image](https://user-images.githubusercontent.com/37487684/55799015-aa65b380-5ad0-11e9-95f1-a7611fde2b40.png)

So i guess there is a probleme, the slice must be renamed to PASSWORD in the GUI above or the ini file modded with USER,PASS like you said, right ?

**Comment by MeowBullet** [comment]
Correct. Whichever you choose... I would probably convert to USER,PASS so all old configs work again without modification.

**Comment by shdf2** [comment]
i think USER,PASS must be the standard. if eveyone customize its environment.ini there will be many problems, because when you share your Configs you'll never remember that people have to change their ini file too...it will not work as expected  and you'll never know why...

**Comment by MeowBullet** [comment]
I don't share them I sell mine, and I have Modified open bullet, in a version I call MeowBullet which is made specifically for my customers, so it will always be USER,PASS as default. But Yes I see your point.

**Comment by shdf2** [comment]
ok, but that's specific to you :) Anyway thank you very much because without any explanation it was (for me) impossible to understand what i was supposed to do...

**Comment by MeowBullet** [comment]
The point is I took measures for mine, you can do the same wether its having a modified version to go with the configs, your own modified environment file zipped with it, or you merely just add a readme to the folder. Lol. It's not hard to let someone know. But yes 99% of people have switched to USER,PASS and I actually do like that it's set as USERNAME,PASSWORD to begin with, kind of keeps the idiots at bay lol. And yes before I read the documentation and opened up the project I was too a bit like what at first.

**Comment by shdf2** [comment]
Well, so actually, all converted configs from .bbc to .loli uses the slice name : PASS
while the environment.ini file says : USERNAME,PASSWORD

Ruri maybe switch to USER,PASS then ? as @MeowBullet said 99% are switching to this. but that 99% is 99% of people aware about it, which is in fact probably only 1% of everyone :)

**Comment by MeowBullet** [comment]
But it's good.... It keeps the idiots a bay

**Comment by openbullet** [comment]
OpenBullet should not be used for credential stuffing so I will not put USER and PASS by default, in order to prevent people who don't know what they are doing to use it for malicious purposes. This was intended, it's not been overlooked.

**Comment by kiviniitty** [comment]
Hello ! anyone know what is my issue here if user:pass contains any symbol like # OB will stuck to just that combo it change just proxy nothing else? I already made some data rules but I dont know what I do wrong 
![Näyttökuva (5)](https://user-images.githubusercontent.com/17779920/56850096-39166500-6906-11e9-9731-0157f23da978.png)
here

**Comment by ghost** [comment]
> Alright I'll see where to put more info about that, but anyways let's say your password slice is called PASSWORD. You will need to add rules like
> `PASSWORD MinLength 8`
> `PASSWORD MaxLength 32`
> `PASSWORD Contains Digit`
> `PASSWORD Contains Lowercase`
> `PASSWORD Contains Uppercase`
> If these are all verified then your password will be accepted as valid and it will be checked with the config, otherwise it will be immediately flagged as a FAIL and OB will move to the next one.
